<center><h1> MCD - Ciencia de Datos Responsable </h1><center>
<center><h2> Tarea 2: AIF360 Toolkit <h2></center>

## Contexto:

El set de datos llamado **"adult dataset"** o **"census income dataset"** contiene datos demográficos y financieros de personas, como su edad, nivel educativo, estado civil, ocupación, relación familiar, entre otros. Su principal finalidad es determinar si una persona percibe un ingreso anual de más o menos de $50,000, siendo así un problema de clasificación binaria. Este conjunto de datos se considera delicado debido a que contiene información personal e ingresos, por lo que regularmente se emplea para enseñar y poner en práctica procesos de análisis de datos de forma ética y responsable.

En la presente tarea se entrega la versión de este dataset disponible en [AIF360](https://aif360.readthedocs.io/en/stable/modules/generated/aif360.datasets.AdultDataset.html#aif360.datasets.AdultDataset), con algunas modificaciones. El objetivo es evaluar la implementación del toolkit AIF360 en el contexto de este dataset, para lo que se entregan una serie de indicaciones.

## Instrucciones:

- La siguiente tarea es **individual**. 
- La calificación de esta tarea se hará de forma **manual**, es decir, esta tarea no cuenta con test cases automáticos. Es por esto que al hacer entrega de su tarea podrá ver una calificación de "0%" y "Aprobado", independiente del contenido de su notebook.
- Cuenta con intentos ilimitados para el envío de su tarea.
- Ante cualquier duda o consulta pueden comunicarse a través de Foros o Salesforce. 

In [112]:
import pandas as pd

In [113]:
df = pd.read_csv('adult_data.csv')
df.head()

age  workclass     education  education-num      marital-status  \
0  25.0    Private          11th            7.0       Never-married   
1  38.0    Private       HS-grad            9.0  Married-civ-spouse   
2  28.0  Local-gov    Assoc-acdm           12.0  Married-civ-spouse   
3  44.0    Private  Some-college           10.0  Married-civ-spouse   
4  34.0    Private          10th            6.0       Never-married   

          occupation   relationship   race   sex  capital-gain  capital-loss  \
0  Machine-op-inspct      Own-child  Black  Male           0.0           0.0   
1    Farming-fishing        Husband  White  Male           0.0           0.0   
2    Protective-serv        Husband  White  Male           0.0           0.0   
3  Machine-op-inspct        Husband  Black  Male        7688.0           0.0   
4      Other-service  Not-in-family  White  Male           0.0           0.0   

   hours-per-week native-country income  
0            40.0  United-States  <=50K  
1            50.0  United-States  <=50K  
2            40.0  United-States   >50K  
3            40.0  United-States   >50K  
4            30.0  United-States  <=50K

### Pregunta 1: Realice el pre-procesamiento de los datos e implemente un modelo de clasificación (Regresión Logística). Obtenga el accuracy del modelo sobre los datos de testeo (1 pto).

#### Indicaciones adicionales:

- Recuerde utilizar random_state o seeds según sea necesario para la reproducibilidad de sus resultados.

### Respuesta

In [114]:
# Seteamos una semilla
seed = 10

In [115]:
# Haremos el pre-procesamiento de los datos
display(df.info())
display(df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45222 entries, 0 to 45221
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             45222 non-null  float64
 1   workclass       45222 non-null  object 
 2   education       45222 non-null  object 
 3   education-num   45222 non-null  float64
 4   marital-status  45222 non-null  object 
 5   occupation      45222 non-null  object 
 6   relationship    45222 non-null  object 
 7   race            45222 non-null  object 
 8   sex             45222 non-null  object 
 9   capital-gain    45222 non-null  float64
 10  capital-loss    45222 non-null  float64
 11  hours-per-week  45222 non-null  float64
 12  native-country  45222 non-null  object 
 13  income          45222 non-null  object 
dtypes: float64(5), object(9)
memory usage: 4.8+ MB


None

age  workclass     education  education-num      marital-status  \
0  25.0    Private          11th            7.0       Never-married   
1  38.0    Private       HS-grad            9.0  Married-civ-spouse   
2  28.0  Local-gov    Assoc-acdm           12.0  Married-civ-spouse   
3  44.0    Private  Some-college           10.0  Married-civ-spouse   
4  34.0    Private          10th            6.0       Never-married   

          occupation   relationship   race   sex  capital-gain  capital-loss  \
0  Machine-op-inspct      Own-child  Black  Male           0.0           0.0   
1    Farming-fishing        Husband  White  Male           0.0           0.0   
2    Protective-serv        Husband  White  Male           0.0           0.0   
3  Machine-op-inspct        Husband  Black  Male        7688.0           0.0   
4      Other-service  Not-in-family  White  Male           0.0           0.0   

   hours-per-week native-country income  
0            40.0  United-States  <=50K  
1            50.0  United-States  <=50K  
2            40.0  United-States   >50K  
3            40.0  United-States   >50K  
4            30.0  United-States  <=50K

In [116]:
# Binarizamos las variables binarias en 0 y 1

df['income'] = df['income'].apply(lambda x: 1 if x == '>50K' else 0)
df['sex'] = df['sex'].apply(lambda x: 1 if x == 'Male' else 0)

In [117]:
# Transformamos a 'category' las columnas 'object'

for col in df.select_dtypes(include='object').columns:
    if df[col].dtype == 'object':
        df[col] = df[col].astype('category')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45222 entries, 0 to 45221
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   age             45222 non-null  float64 
 1   workclass       45222 non-null  category
 2   education       45222 non-null  category
 3   education-num   45222 non-null  float64 
 4   marital-status  45222 non-null  category
 5   occupation      45222 non-null  category
 6   relationship    45222 non-null  category
 7   race            45222 non-null  category
 8   sex             45222 non-null  int64   
 9   capital-gain    45222 non-null  float64 
 10  capital-loss    45222 non-null  float64 
 11  hours-per-week  45222 non-null  float64 
 12  native-country  45222 non-null  category
 13  income          45222 non-null  int64   
dtypes: category(7), float64(5), int64(2)
memory usage: 2.7 MB


In [118]:
# Separamos X e Y.
target = 'income'

X = df.loc[:, df.columns != target]
y = df.loc[:, df.columns == target]

In [119]:
# Obtenemos variables dummies

df1 = X.copy()

ignore = ['sex', target]
for catcol in df1.select_dtypes(include='category').columns:
    if catcol in ignore:
      pass
    else:
      dummies = pd.get_dummies(X[catcol])
      dummies.columns = [catcol+ '_' + x for x in dummies.columns]
      X = pd.concat([X, dummies], axis=1).drop(columns=[catcol])

X.head()

age  education-num  sex  capital-gain  capital-loss  hours-per-week  \
0  25.0            7.0    1           0.0           0.0            40.0   
1  38.0            9.0    1           0.0           0.0            50.0   
2  28.0           12.0    1           0.0           0.0            40.0   
3  44.0           10.0    1        7688.0           0.0            40.0   
4  34.0            6.0    1           0.0           0.0            30.0   

   workclass_Federal-gov  workclass_Local-gov  workclass_Private  \
0                      0                    0                  1   
1                      0                    0                  1   
2                      0                    1                  0   
3                      0                    0                  1   
4                      0                    0                  1   

   workclass_Self-emp-inc  ...  native-country_Portugal  \
0                       0  ...                        0   
1                       0  ...                        0   
2                       0  ...                        0   
3                       0  ...                        0   
4                       0  ...                        0   

   native-country_Puerto-Rico  native-country_Scotland  native-country_South  \
0                           0                        0                     0   
1                           0                        0                     0   
2                           0                        0                     0   
3                           0                        0                     0   
4                           0                        0                     0   

   native-country_Taiwan  native-country_Thailand  \
0                      0                        0   
1                      0                        0   
2                      0                        0   
3                      0                        0   
4                      0                        0   

   native-country_Trinadad&Tobago  native-country_United-States  \
0                               0                             1   
1                               0                             1   
2                               0                             1   
3                               0                             1   
4                               0                             1   

   native-country_Vietnam  native-country_Yugoslavia  
0                       0                          0  
1                       0                          0  
2                       0                          0  
3                       0                          0  
4                       0                          0  

[5 rows x 102 columns]

In [120]:
# Datos de entrenamiento y prueba
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=908)

In [121]:
# Construimos la regresion logistica
import numpy as np
from sklearn.linear_model import LogisticRegressionCV

lr = LogisticRegressionCV(solver='liblinear', cv=10, random_state=seed)
lr.fit(X_train, np.ravel(y_train))

LogisticRegressionCV(cv=10, random_state=10, solver='liblinear')

In [122]:
# Obtenemos el accuracy sobre los datos de testeo
from sklearn.metrics import accuracy_score


y_pred_proba = lr.predict_proba(X_test)[:,1]
y_pred = y_pred_proba >= 0.5

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy}")


Accuracy: 0.8475400773908237


### Pregunta 2: Evalúe las clasificaciones del modelo de la pregunta anterior a través de las métricas de Fairness facilitadas en AIF360 utilizando la raza o sexo como atributo protegido (elija una). Destaque alguna métrica de fairness que le sea de interés (1 pto).

#### Indicaciones adicionales:

- Obtenga tanto las intancias de BinaryLabelDatasetMetric como de ClassificationMetric y calcule al menos una métrica de Fairness.

- Considere 'Black' como grupo no privilegiado para la raza, 'White' como grupo privilegiado para la raza.
- Considere 'Female' como grupo no privilegiado para el sexo, 'Male' como grupo privilegiado para el sexo.

- Está invitado(a) a realizar el análisis para ambos atributos sensibles, aunque **no es obligatorio**.

### Respuesta

In [146]:
# Protegeremos el atributo sexo como atributo protegido
privileged_group = [{'sex': 1}]
unprivileged_group = [{'sex': 0}]

In [148]:
# Preparamos el dataset de test

df_test = X_test.copy()
df_test[target] = y_test

df_test

age  education-num  sex  capital-gain  capital-loss  hours-per-week  \
24938  23.0           13.0    1           0.0           0.0            30.0   
22771  35.0           13.0    1        4386.0           0.0            47.0   
6957   45.0           14.0    0           0.0           0.0            45.0   
23546  28.0            9.0    1           0.0           0.0            40.0   
18142  41.0            9.0    1           0.0           0.0            40.0   
...     ...            ...  ...           ...           ...             ...   
29934  36.0           10.0    0           0.0           0.0            40.0   
10608  35.0            9.0    1           0.0           0.0            40.0   
32133  39.0            5.0    1           0.0           0.0            35.0   
41809  48.0            9.0    1           0.0           0.0            45.0   
18705  41.0           10.0    0           0.0           0.0            38.0   

       workclass_Federal-gov  workclass_Local-gov  workclass_Private  \
24938                      0                    0                  1   
22771                      0                    0                  1   
6957                       0                    1                  0   
23546                      0                    0                  1   
18142                      0                    0                  1   
...                      ...                  ...                ...   
29934                      0                    0                  1   
10608                      0                    0                  1   
32133                      0                    0                  1   
41809                      0                    0                  1   
18705                      0                    0                  1   

       workclass_Self-emp-inc  ...  native-country_Puerto-Rico  \
24938                       0  ...                           0   
22771                       0  ...                           0   
6957                        0  ...                           0   
23546                       0  ...                           0   
18142                       0  ...                           0   
...                       ...  ...                         ...   
29934                       0  ...                           0   
10608                       0  ...                           0   
32133                       0  ...                           0   
41809                       0  ...                           0   
18705                       0  ...                           0   

       native-country_Scotland  native-country_South  native-country_Taiwan  \
24938                        0                     0                      0   
22771                        0                     0                      0   
6957                         0                     0                      0   
23546                        0                     0                      0   
18142                        0                     0                      0   
...                        ...                   ...                    ...   
29934                        0                     0                      0   
10608                        0                     0                      0   
32133                        0                     0                      0   
41809                        0                     0                      0   
18705                        0                     0                      0   

       native-country_Thailand  native-country_Trinadad&Tobago  \
24938                        0                               0   
22771                        0                               0   
6957                         0                               0   
23546                        0                               0   
18142                        0                               0   
...                        ...                          

In [149]:
# Instanciamos el dataset de test por medio de BinaryLabelDataset
from aif360.datasets import BinaryLabelDataset

attributes_params = dict(protected_attribute_names=['sex'], label_names=[target])
dt_test = BinaryLabelDataset(df=df_test, **attributes_params,
                              favorable_label=1, unfavorable_label=0)

In [154]:
# Instanciamos la metrica de test por medio de BinaryLabelDatasetMetric
from aif360.metrics import BinaryLabelDatasetMetric

metric_sex_test_df = BinaryLabelDatasetMetric(dt_test,
                                        unprivileged_groups=unprivileged_group,
                                        privileged_groups=privileged_group)

In [155]:
# Utilizamos las inferencias anteriores
df_pred = X_test.copy()
df_pred[target] = y_pred

In [156]:
# Repetimos el proceso de calcular el dataset según los grupos privilegiados y no privilegiados

dt_pred = BinaryLabelDataset(df=df_pred, **attributes_params,
                             favorable_label=1, unfavorable_label=0)
metric_sex_pred_df = BinaryLabelDatasetMetric(dt_pred,
                                            unprivileged_groups=unprivileged_group,
                                            privileged_groups=privileged_group)

In [153]:
# Utilizando ClassificationMetric, calculamos la diferencia entre las predicciones y los datos de entrenamiento
from aif360.metrics import ClassificationMetric

pred_vs_test_classification_metric = ClassificationMetric(dt_test, dt_pred,
                                                    unprivileged_groups=unprivileged_group,
                                                    privileged_groups=privileged_group)

In [157]:
# Calculamos una métrica de fairness, en este caso 'Statistical Parity Difference'

# Para el dataset de test
print(f'Test Statistical Parity Difference (sex): {metric_sex_test_df.mean_difference()}')

Test Statistical Parity Difference (sex): -0.20439336098065602


Vemos que para los datos de test, tenemos que la métrica de diferencia estadística es negativa. Esto significa que la probabilidad de pertenecer a la clase positiva para el sexo femenino es menor que para el sexo masculino, lo que indica presencia de sesgo.

In [158]:
# Replicamos lo mismo pero para las predicciones
print(f'Pred Statistical Parity Difference (sex): {metric_sex_pred_df.mean_difference()}')

Pred Statistical Parity Difference (sex): -0.16910030241166435


Similar a los datos de test, tenemos que la métrica de diferencia estadística es negativa para nuestras predicciones, lo que aún indica sesgo en el modelo entrenado y en los datos, confirmado porque el modelo fue entrenado con datos sesgados, por ende las predicciones también repetirán el patrón.

In [167]:
# Ahora comparamos el ratio de error entre los datos de test y las predicciones
print(f'Error rate difference (unprivileged error rate - privileged error rate)= {pred_vs_test_classification_metric.error_rate_difference()}\n')

Error rate difference (unprivileged error rate - privileged error rate)= -0.12029179216293007



Por otra parte, tenemos que la diferencia de la tasa de error entre ambos dataset es negativa. Esto indica que el ratio de error para el grupo no privilegiado (sexo femenino) es menor que para el grupo privilegiado (sexo masculino). En otras palabras, el modelo tiene mejores predicciones para el sexo femenino vs el sexo masculino, teniendo una tendencia de clasificar con más precisión al sexo femenino en el grupo de bajos ingresos.

### Pregunta 3: Aplique el algoritmo EqOddsPostProcessing sobre las predicciones del modelo de la pregunta 1. Obtenga las predicciones modificadas y evalúelas. ¿Qué cambios hubo en las métricas de Fairness? (2 ptos). 

#### Indicaciones adicionales:

- Obtenga tanto las intancias de BinaryLabelDatasetMetric como de ClassificationMetric y calcule al menos una métrica de Fairness.


### Respuesta

In [161]:
# Utilizamos EqOddsPostprocessing como post procesamiento para cambiar las predicciones, mejorando la tasa de verdaderos positivos
# y falsos positivos
from aif360.algorithms.postprocessing.eq_odds_postprocessing import EqOddsPostprocessing


roc = EqOddsPostprocessing(unprivileged_groups=unprivileged_group,
                            privileged_groups=privileged_group,
                            seed = seed) # Utilizamos la misma métrica

roc.fit(dt_test, dt_pred)

In [168]:
# Obtenemos las predicciones y evaluamos bajo las mismas métricas de Fairness
df_pred_eopp = roc.predict(dt_pred)

metric_sex_pred_eopp_df = BinaryLabelDatasetMetric(df_pred_eopp,
                                  unprivileged_groups=unprivileged_group,
                                  privileged_groups=privileged_group)

pred_vs_test_classification_metric_eopp = ClassificationMetric(dt_test, df_pred_eopp,
                                                    unprivileged_groups=unprivileged_group,
                                                    privileged_groups=privileged_group)

In [169]:
# Calculamos 'Statistical Parity Difference' bajo las predicciones modificados con EqOddsPostprocessing

# Para el dataset de test
print(f'Pred EdOddsPostProcessing Statistical Parity Difference (sex): {metric_sex_pred_eopp_df.mean_difference()}')

Pred EdOddsPostProcessing Statistical Parity Difference (sex): -0.10341197257239113


Vemos que efectivamente con el método de post procesamiento nos hemos acercado al valor 0 de la diferencia estadística, por lo que hemos disminuido el sesgo en contra del grupo no privilegiado (sexo femenino) en torno a las predicciones obtenidas.

In [170]:
# Ahora comparamos el ratio de error entre los datos de test y las predicciones modificadas
print(f'Error rate difference (unprivileged error rate - privileged error rate)= {pred_vs_test_classification_metric_eopp.error_rate_difference()}\n')

Error rate difference (unprivileged error rate - privileged error rate)= -0.06229039453889085



Similar a la tendencia de arriba, hemos disminuido a la mitad la diferencia de tasas de error entre el grupo no privilegiado y el grupo privilegiado. Esto puede indicar que la modificación de las predicciones mejora la justicia de aquellas con respecto al sexo femenino, en el que las tasas de errores de tanto este grupo como el grupo privilegiado (sexo masculino) son más similares en este momento.

### Pregunta 4: Aplique el algoritmo de Reweighing sobre los datos de entrenamiento. Compare las métricas de Fairness para los datos de entrenamiento antes y después de aplicar Reweighing. Luego, entrene un modelo de Regresión Logística utilizando los datos de entrenamiento modificados y obtenga las métricas de Fairness de las predicciones de dicho modelo. ¿Cómo varió el accuracy del modelo respecto al original y las métricas de fairness? (2 ptos).

#### Indicaciones adicionales:

- Obtenga tanto las intancias de BinaryLabelDatasetMetric como de ClassificationMetric y calcule al menos una métrica de Fairness.

- Considere 'Black' como grupo no privilegiado para la raza, 'White' como grupo privilegiado para la raza.
- Considere 'Female' como grupo no privilegiado para el sexo, 'Male' como grupo privilegiado para el sexo.

- Nuevamente, elija raza o sexo para el cálculo de las métricas. Está invitado(a) a realizar el análisis para ambos atributos sensibles, aunque **no es obligatorio**.

### Respuesta

In [172]:
# Ahora aplicaremos cambios en los datos de entrenamiento para disminuir el sesgo entre grupos privilegiados y no privilegiados

# Preparamos el dataset de entrenamiento
df_train = X_train.copy()
df_train[target] = y_train

dt_train = BinaryLabelDataset(df=df_train, **attributes_params,
                             favorable_label=1, unfavorable_label=0)
metric_sex_train_df = BinaryLabelDatasetMetric(dt_train,
                                            unprivileged_groups=unprivileged_group,
                                            privileged_groups=privileged_group)

In [179]:
# Vemos los pesos iniciales para el dataset de entrenamiento. Notamos que todos están seteados en 1
for i in dt_train.instance_weights:
    if i != 1:
        print(i)

In [180]:
# Instanciamos algoritmo de Reweighing
from aif360.algorithms.preprocessing import Reweighing


reweight = Reweighing(unprivileged_groups=unprivileged_group,
                privileged_groups=privileged_group)

In [184]:
# Obtenemos los nuevos datos de entrenamiento con nuevos pesos
reweighted = reweight.fit_transform(dt_train)
reweighted_sex_train_df = BinaryLabelDatasetMetric(reweighted,
                                              unprivileged_groups=unprivileged_group,
                                              privileged_groups=privileged_group)

In [182]:
# Vemos los pesos iniciales ahora cambiaron para el dataset de entrenamiento
for i in reweighted.instance_weights:
    if i != 1:
        print(i)

0.8497495760182721
1.0940277426196912
0.7933168432064275
2.1551190818742096
0.7933168432064275
0.7933168432064275
1.0940277426196912
0.7933168432064275
1.0940277426196912
1.0940277426196912
1.0940277426196912
0.8497495760182721
0.8497495760182721
0.7933168432064275
1.0940277426196912
1.0940277426196912
0.7933168432064275
1.0940277426196912
1.0940277426196912
2.1551190818742096
0.8497495760182721
0.7933168432064275
0.7933168432064275
0.7933168432064275
1.0940277426196912
1.0940277426196912
0.7933168432064275
1.0940277426196912
1.0940277426196912
0.7933168432064275
0.8497495760182721
1.0940277426196912
1.0940277426196912
2.1551190818742096
1.0940277426196912
1.0940277426196912
0.7933168432064275
1.0940277426196912
2.1551190818742096
1.0940277426196912
1.0940277426196912
1.0940277426196912
1.0940277426196912
1.0940277426196912
0.8497495760182721
1.0940277426196912
2.1551190818742096
0.7933168432064275
0.8497495760182721
1.0940277426196912
0.7933168432064275
0.8497495760182721
1.0940277426

In [185]:
# Comparamos la diferencia de 'Statistical Parity Difference' entre los datos de entrenamiento y los datos modificados
print(f'Training Statistical Parity Difference (sex): {metric_sex_train_df.mean_difference()}')
print(f'Reweighted Statistical Parity Difference (sex): {reweighted_sex_train_df.mean_difference()}')

Training Statistical Parity Difference (sex): -0.1975830166220287
Reweighted Statistical Parity Difference (sex): -2.7755575615628914e-17


Vemos una diferencia notable entre el dataset de entrenamiento con pesos iniciales en 1, versus con pesos preprocesados. Después del pre procesamiento, prácticamente no hay diferencia estadística entre ambos grupos (sexo femenino vs sexo masculino).

In [186]:
# Ahora entrenamos una regresión logística con los datos de entrenamiento, teniendo los pesos modificados
lr_rw = LogisticRegressionCV(solver='liblinear', cv=10, random_state=seed)
lr_rw.fit(X_train, np.ravel(y_train), sample_weight=reweighted.instance_weights)

LogisticRegressionCV(cv=10, random_state=10, solver='liblinear')

In [188]:
# Calculamos el nuevo accuracy y comparamos con el anterior
y_pred_proba_rw = lr_rw.predict_proba(X_test)[:,1]
y_pred_rw = y_pred_proba_rw >= 0.5
accuracy_rw = accuracy_score(y_test, y_pred_rw)

print(f"Accuracy normal: {accuracy}")
print(f"Accuracy reweighted: {accuracy_rw}")

Accuracy normal: 0.8475400773908237
Accuracy reweighted: 0.8403537866224433


En este caso, no vemos cambios significativos en el accuracy. De hecho, se ve que ligeramente el accuracy con nuevos pesos es ligeramente menor. Ahora mediremos métricas de Fairness.

In [189]:
# Repetimos el proceso de calcular el dataset según los grupos privilegiados y no privilegiados pero ahora con las
# predicciones desde los datos con nuevos pesos

df_pred_rw = X_test.copy()
df_pred_rw[target] = y_pred_rw


dt_pred_rw = BinaryLabelDataset(df=df_pred_rw, **attributes_params,
                             favorable_label=1, unfavorable_label=0)
metric_sex_pred_rw_df = BinaryLabelDatasetMetric(dt_pred_rw,
                                            unprivileged_groups=unprivileged_group,
                                            privileged_groups=privileged_group)

In [190]:
# Comparamos la diferencia de 'Statistical Parity Difference' entre las predicciones originales y 
# las predicciones con el modelo entrenado bajo nuevos pesos

print(f'Pred Original Statistical Parity Difference (sex): {metric_sex_pred_df.mean_difference()}')
print(f'Pred Reweighted Statistical Parity Difference (sex): {metric_sex_pred_rw_df.mean_difference()}')

Pred Original Statistical Parity Difference (sex): -0.16910030241166435
Pred Reweighted Statistical Parity Difference (sex): -0.07497581622159358


Es muy notable la diferencia nuevamente con el pre procesamiento de los pesos, para el caso de las predicciones. Podemos ver que repesar los atributos bajo un grupo no privilegiado genera predicciones con menor sesgo y mejor Fairness, dado que la métrica estadística es más cercana a 0.

In [197]:
# Finalmente, comparamos el ratio de error entre los datos de tests y las predicciones con los nuevos pesos

test_vs_pred_rw_classification_metric = ClassificationMetric(dt_test, dt_pred_rw,
                                                    unprivileged_groups=unprivileged_group,
                                                    privileged_groups=privileged_group)

print(f'Error rate difference before (unprivileged error rate - privileged error rate)= {pred_vs_test_classification_metric.error_rate_difference()}')
print(f'Error rate difference now (unprivileged error rate - privileged error rate)= {test_vs_pred_rw_classification_metric.error_rate_difference()}\n')

Error rate difference before (unprivileged error rate - privileged error rate)= -0.12029179216293007
Error rate difference now (unprivileged error rate - privileged error rate)= -0.10166951989478146



Nuevamente, notamos que la tasa de errores entre el dataset de test y las nuevas predicciones se acerca más a 0. Anteriormente estábamos ante una diferencia de errores en `-0.12` y ahora mejoramos a `-0.10`. Lo anterior también permite validar que en las predicciones dado el pre procesamiento de pesos se disminuye el sesgo de los datos. Importante considerar que esto es consistente con lo que ocurre en datos de entrenamiento: cuando los datos que ingresan a un modelo tienen un cierto sesgo, muy probablemente las predicciones tenderán a repetir este patrón de igual manera. Dado lo anterior, vale la pena aplicar pre y post procesamiento en torno a Fairness, ya que pese a que el `accuracy` puede mantenerse similar, el sesgo puede disminuir correctamente, haciendo un mejor uso de elementos de IA que sean más beneficiosos para la sociedad.